docs/proposals/20210310-edge-device-management.md

```diff
@@ -0,0 +1,498 @@
+      "modified": <modifed-timestamp>,
+      "id": "<command-id>",
+      "name": "lightcolor",
+      "get": {
+        "path": "/api/v1/device/{deviceId}/color",
+        "responses": [
+          {
+            "code": "200",
+            "description": "get current light color",
+            "expectedValues": [
+              "color"
+            ]
+          },
+          {
+            "code": "503",
+            "description": "service unavailable"
+          }
+        ],
+        "url": "http://edgex-core-command:48082/api/v1/device/<device-id>/command/<set-command-id>"
+      },
+      "put": {
+        "path": "/api/v1/device/{deviceId}/changeColor",
+        "responses": [
+          {
+            "code": "201",
+            "description": "set the light color"
+          },
+          {
+            "code": "503",
+            "description": "service unavailable"
+          }
+        ],
+        "url": "http://edgex-core-command:48082/api/v1/device/<device-id>/command/<put-command-id>"
+      }
+    }
+  ]
+}
+```
+
+Then we can create a `Device` associated to the `DeviceProfile` as following,
+
+```yaml
+apiVersion: device.openyurt.io/v1alpha1
+kind: Device
+metadata:
+  name: testsensor
+spec:
+  ...
+  profile: color-sensor
+  deviceProperties:
+    lightcolor:
+      name: lightcolor
+      desiredValue: green
+```
+
+Assume that the initial light color is blue, then after applying the `Device`, the device controller will try to fetch the corresponding get/put URLs from the core command service, update the `testsensor.Spec.DeviceProperties[lightcolor]` `testsensor.Status.DeviceProperties[lightcolor]`, and then set the light color to green, the updated `Device` will look like the following,
+```yaml
+apiVersion: device.openyurt.io/v1alpha1
+kind: Device
+metadata:
+  name: testsensor
+spec:
+  ...
+  profile: color-sensor
+  deviceProperties:
+    lightcolor:
+      name: lightcolor
+      setURL: http://edgex-core-command:48082/api/v1/device/<device-id>/command/<put-command-id>
+      desiredValue: green
+status:
+  deviceProperties:
+    lightcolor:
+      name: lightcolor
+      getURL: http://edgex-core-command:48082/api/v1/device/<device-id>/command/<set-command-id>
+      desiredValue: green
+```
+
+### System Architecture
+
+Edge devices are usually located in the network regions that are unreachable from the cloud, while `DeviceService` needs to talk to edge devices directly. Therefore, we need to deploy the `DeviceService` on the edge node that connects to edge devices. On the other hand, when modifying the device properties, the device controller will need to send a request to the `DeviceService`. Hence, we will deploy a replica of the device controller in each network region. Specifically, the device controller will connect to the YurtHub, and the YurtHub will only pull the device CRDs related to connected edge devices. The system architecture is shown below,
+
+![system-architecture](../img/system-arch.png)
```

Comment:
We should consider follow scenario:
EdgeX per network region:  
If we only deploy the Edgex on the cloud then we have an assumption that the network connection between the edge and cloud is quite stable. Since the device service need to push the sensor data to the core data service for persistent and later processing. If we only allow deploy the Edgex on the cloud we should expect that there are data losing during the network breakdown period. 

Beside that deploy the Edgex  near/on the Edge will shorten the latency.  Support multiple EdgeX instances will make it easier for scale out and information isolation.

Reply:
Yes. As we discussed offline, we will add a new CRD, EdgeXService (the name can be changed later), which will define the way the EdgeX service is deployed. Also when initiating the EdgeX service, we will synchronize all existing objects (DeviceProfile, DeviceService and Device) on the EdgeX to the OpenYurt.